AikumaCloudStorage/src/org/lp20/aikuma/storage/GoogleDriveFolderCache.java

```diff
@@ -0,0 +1,178 @@
+ * should be called after all tuples have been added. This approach requires
+ * a little care regarding thread synchronization. Because beginTable()
+ * creates a lock, finishTable() must be called to unlock.
+ *
+ * Another way of building the cache is to explicitly add path and Google file
+ * ID pairs. Use add() to add a pair. Use getFid() and getPath() to get Google
+ * File ID and directory path respectively.
+ */
+class GoogleDriveFolderCache {
+	final Map<String,String> mFids = new HashMap<String,String>();
+	final Map<String,String> mPaths = new HashMap<String,String>();
+	Map<String,String> mNames;
+	Map<String,Set<String>> mPids0;
+	Map<String,String> mPids;
+	private final ReentrantLock lock = new ReentrantLock();
+	private static GoogleDriveFolderCache mObj = null;
+
+	private GoogleDriveFolderCache() {
+	}
+
+	/**
+	 * Get the singletone instance of the class.
+	 */
+	public static synchronized GoogleDriveFolderCache getInstance() {
+		if (mObj == null)
+			mObj = new GoogleDriveFolderCache();
+		return mObj;
+	}
+
+	/**
+	 * This exception signals a problem in building the cache.
+	 */
+	public class Error extends Exception {
+		public Error(String message) {
+			super(message);
+		}
+	}
+
+	private String computePath(String id) throws Error {
+		if (mPaths.containsKey(id)) {
+			return mPaths.get(id);
+		}
+		
+		String path;
+		if (mPids.containsKey(id)) {
+			String prefix = computePath(mPids.get(id));
+                        String joiner = prefix.equals("/") ? "" : "/";
+                        path = prefix + joiner + mNames.get(id);
+		} else {
+			path = "/";
+		}
+
+		mPaths.put(id, path);
+		if (mFids.containsKey(path)) {
+			String msg = String.format("folder maps to multiple paths: %s (%s)", path, id);
+			throw new Error(msg);
+		}
+		mFids.put(path, id);
+		return path;
+	}
+
+	/**
+	 * Start building the cache.
+	 *
+	 * This is one way of building the cache. After this, use addToTable() to
+	 * add materials to build the cache, and then call finishTable() to finalize
+	 * building the cache.
+	 */
+	public void beginTable() {
+		lock.lock();
+		mNames = new HashMap<String,String>();
+		mPids0 = new HashMap<String,Set<String>>();
+		mPids = new HashMap<String,String>();
+	}
+
+	/**
+	 * Add information for a directory which is later used to build the cache.
+	 *
+	 * @param id Google file ID of the directory.
+	 * @param title Name of the directory.
+	 * @param pid Google file ID of the parent directory.
+	 */
+	public void addToTable(String id, String title, String pid) {
+		mNames.put(id, title);
+		if (!mPids0.containsKey(id))
+			mPids0.put(id, new HashSet<String>());
+		mPids0.get(id).add(pid);
+	}
+
+	/**
+	 * Build the cache using the information added by addToTable().
+	 */
+	public void finishTable() throws Error {
+		for (String id: mPids0.keySet()) {
+			for (String pid: mPids0.get(id)) {
+				if (mNames.containsKey(pid)) {
+					mPids.put(id, pid);
```

Comment:
A folder can have multiple parent folders. If the folder is inside the aikuma dir tree and has multiple parents, one and only one of them should be inside the aikuma dir tree. The rest of them should be outside the tree. (`computePath` detects violation of this assumption.) This loop makes sure to pick the right parent, the one inside the aikuma dir tree. Because `mNames` is basically a list of all nodes in the aikuma dir tree, we use it to tell whether a given parent node is inside the aikuma dir tree.